internal/civisibility/integrations/civisibility_features.go

```diff
@@ -0,0 +1,108 @@
+	additionalFeaturesInitializationOnce sync.Once
+
+	// ciVisibilityRapidClient contains the http rapid client to do CI Visibility querys and upload to the rapid backend
+	ciVisibilityClient net.Client
+
+	// ciVisibilitySettings contains the CI Visibility settings for this session
+	ciVisibilitySettings net.SettingsResponseData
+
+	// ciVisibilityEarlyFlakyDetectionSettings contains the CI Visibility Early Flake Detection data for this session
+	ciVisibilityEarlyFlakyDetectionSettings net.EfdResponseData
+
+	// ciVisibilityFlakyRetriesSettings contains the CI Visibility Flaky Retries settings for this session
+	ciVisibilityFlakyRetriesSettings FlakyRetriesSetting
+)
+
+// ensureAdditionalFeaturesInitialization initialize all the additional features
+func ensureAdditionalFeaturesInitialization(serviceName string) {
+	additionalFeaturesInitializationOnce.Do(func() {
+		// Create the CI Visibility client
+		ciVisibilityClient = net.NewClientWithServiceName(serviceName)
+		if ciVisibilityClient == nil {
+			log.Error("CI Visibility: error getting the ci visibility http client")
+			return
+		}
+
+		// Get the CI Visibility settings payload for this test session
+		ciSettings, err := ciVisibilityClient.GetSettings()
+		if err != nil {
+			log.Error("CI Visibility: error getting CI visibility settings: %v", err)
+		} else if ciSettings != nil {
+			ciVisibilitySettings = *ciSettings
+		}
+
+		// if early flake detection is enabled then we run the early flake detection request
+		if ciVisibilitySettings.EarlyFlakeDetection.Enabled {
+			ciEfdData, err := ciVisibilityClient.GetEarlyFlakeDetectionData()
+			if err != nil {
+				log.Error("CI Visibility: error getting CI visibility early flake detection data: %v", err)
+			} else if ciEfdData != nil {
+				ciVisibilityEarlyFlakyDetectionSettings = *ciEfdData
+			}
+		}
+
+		// if flaky test retries is enabled then let's load the flaky retries settings
+		if ciVisibilitySettings.FlakyTestRetriesEnabled {
+			flakyRetryEnabledByEnv := internal.BoolEnv(constants.CIVisibilityFlakyRetryEnabledEnvironmentVariable, true)
+			if flakyRetryEnabledByEnv {
+				totalRetriesCount := (int64)(internal.IntEnv(constants.CIVisibilityTotalFlakyRetryCountEnvironmentVariable, 1_000))
+				ciVisibilityFlakyRetriesSettings = FlakyRetriesSetting{
+					RetryCount:               (int64)(internal.IntEnv(constants.CIVisibilityFlakyRetryCountEnvironmentVariable, 5)),
+					TotalRetryCount:          totalRetriesCount,
+					RemainingTotalRetryCount: totalRetriesCount,
+				}
+			} else {
+				log.Warn("CI Visibility: flaky test retries was disabled by the environment variable")
+				ciVisibilitySettings.FlakyTestRetriesEnabled = false
+			}
+		}
+	})
+}
+
+// GetSettings gets the settings from the backend settings endpoint
+func GetSettings() *net.SettingsResponseData {
+	// call to ensure the additional features initialization is completed (service name can be null here)
```

Comment:
If this happens to be called before ` ensureAdditionalFeaturesInitialization()` with a non-empty `serviceName` is called, the initialization will be performed with an empty `serviceName` and subsequent calls will do nothing, right? Could this be a problem?

Reply:
if `serviceName` is empty, then we recalculate the serviceName in the http client, do you remember the regex bug that you spotted? 😄  That one recalculate the serviceName.
That means that sending the serviceName is just a small optimization to avoid calculating the serviceName twice.